New package: vosk-api-0.3.43

JohnGebbie opened the pull request:
#### Testing the changes
- I tested the changes in this PR: **YES**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

#### Local build testing
- I built this PR locally for my native architecture, (x86_64-glibc)

Closes https://github.com/void-linux/void-packages/issues/39009

This speech recognition library is a dependency of my assistive voice input program (https://github.com/void-linux/void-packages/pull/39716).
The build requires building their forked version of Kaldi which has different dependencies and a different process from the upstream.
I've patched it to use the latest version of openblas that's already packaged, and the latest version of openfst which I've also committed a template for.

JohnGebbie commented:
I'm guessing my template's no good, and I'm trying to write one that does all the compiling.

Vosk has a bunch of dependencies that have been modified for itself and wouldn't make packages of their own.
How do I write a template with multiple things to build?
What happens with multiple distfiles, when I `ls` in the template, it just seems to be the files of one of them?
Can you have multiple buid_styles?

JohnGebbie commented:
I think that's it there, except I'll try to update it to use the openblas package.

JohnGebbie commented:
That's it ready.

JohnGebbie commented:
That's it ready.

JohnGebbie commented:
I'll work on the other architectures but do you know why the top one failed at the "Compare to previous" step?

JohnGebbie commented:
That's it working with the `openblas-devel` package. I still need to work on the other architectures.

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+pkgname=vosk-api
+version=0.3.43
+revision=1
+_openblas_ver=0.3.20
+_clapack_ver=3.2.1
+_openfst_commit=7dfd808194105162f20084bb4d8e4ee4b65266d5
+_kaldi_commit=6417ac1dece94783e80dfbac0148604685d27579
```
could it work with system packages?

JohnGebbie replied:
Thanks paper42.
I'd forgotten to remove the openblas and clapack variables, which have been replaced with system packages.
The version of kaldi is vosk specific so I think it's best as it is.
I tried packaging openfst separately but the newer version wasn't compatible (and there's vosk specific changes but I didn't get that far).

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+patch_args="-Np1 ../kaldi-${_kaldi_commit}/src/matrix/cblas-wrappers.h"
+
+post_extract() {
+	cd "${XBPS_BUILDDIR}/vosk-api-${version}/src"
```
use create_wrksrc to contain all of the extracted distfiles in one directory and fix cleaning

JohnGebbie commented:
It works for x86_64, but for aarch64 Kaldi errors saying there's no libopenblas.so.

JohnGebbie commented:
I think that's it :)

JohnGebbie commented:
Added do_check().

JohnGebbie commented:
That should work.

JohnGebbie commented:
I think it's ready. I've tested x86_64, i686 and aarch64-musl.

JohnGebbie commented:
just git pulling it up to date.

JohnGebbie commented:
I added the compile options the i686 ci mentioned it's error message, and it worked on my machine but it did before too. I also added a fix in do_check for x86_64, I didn't clean and recompile but `./xbps-src check` worked, so I hope it works.

JohnGebbie commented:
Pushed a fix. I think the problem was the lack of --build and maybe why it only worked on my machine. openfst piled for i686 and x86_64.

JohnGebbie commented:
Could i get a run of i686 please?

JohnGebbie commented:
I added --host="${XBPS_TRIPLET}" for when not cross compiling. i686 is still cross compiling successfully on my machine. I also git pulled to to update my branch.

JohnGebbie commented:
I pushed stopping it running tests in the configure phase when natively compiling and added -lexecinfo in do_check for musl.

JohnGebbie commented:
wait I made a mistake.

JohnGebbie commented:
Ok hopefully it passes the ci. i686 cross compiled on mine.

JohnGebbie commented:
Could we try the ci again? seems there was just a problem fetching a distfile.

JohnGebbie commented:
Could I get the tests run please

JohnGebbie commented:
just git pulled my branch up to date, still ready

JohnGebbie commented:
I think I should have added the .so to common/shlibs.

JohnGebbie commented:
All I just did was update the branch and add the SONAME to common/shlibs.

JohnGebbie commented:
Still a work in progress, for some reason `./xbps-src show-shlib-provides vosk-api` prints nothing.

JohnGebbie commented:
The reason was `objdump -p libvosk.so` doesn't have SONAME so the hook ignores it, so I've added `shlib_provides="libvosk.so"`. The package should be ready, I'll try to use it for my program and report back.

JohnGebbie commented:
I've tested this with my program https://github.com/void-linux/void-packages/pull/39716 and it's working well (I'm using it to type this!). I'm happy with the template and think it's ready.

JohnGebbie commented:
Is there hope? I've tried to make it as nice as possible and I've got it using as many system packages for the dependencies as I could.

JohnGebbie commented:
I'm going to try again to get it working with the latest version of openfst, so openfst could be packaged separately.

JohnGebbie commented:
That's it ready to merge

JohnGebbie commented:
what do you think @paper42?
here's a video of me using it: https://numen.johngebbie.com

paper42 commented on srcpkgs/openfst/template:
```diff
@@ -0,0 +1,59 @@
+distfiles="https://www.openfst.org/twiki/pub/FST/FstDownload/openfst-$version.tar.gz"
+checksum=de987bf3624721c5d5ba321af95751898e4f4bb41c8a36e2d64f0627656d8b42
+
+post_extract() {
+	# Don't run test during build.
+	vsed -i configure.ac -e '152,184 d'
+	# Remove -msse check confused by cross compiling.
+	vsed -i src/include/fst/float-weight.h -e '120 d'
+}
```
specifying which lines should be deleted is very fragile, I think a patch would be better.

JohnGebbie replied:
Thank you so much, made me very happy seeing you'd reviewed!

I added a patch to allow cross compiling, but a test about float equality fails when natively compiling i686, so I have limited the template to x86_64* and aarch64* with a comment to the issue I've opened, until it is resolved.

I see they're doing the same on alpine: https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/41097

paper42 commented on srcpkgs/openfst/template:
```diff
@@ -0,0 +1,59 @@
+
+openfst-devel_package() {
+	depends="${sourcepkg}>=${version}_${revision}"
+	short_desc+=" -- development files"
```
```suggestion
	short_desc+=" - development files"
```

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,79 @@
+revision=1
+create_wrksrc=yes
+build_wrksrc="vosk-api-${version}/src"
+_openfst_commit=7dfd808194105162f20084bb4d8e4ee4b65266d5
```
I don't see this used anywhere, is this here from when openfst was vendored in this package?

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,79 @@
+fi
+
+post_extract() {
+	# Don't need Fortran to C, we have lapack not clapack.
+	cd "${wrksrc}/kaldi-${_kaldi_commit}/src"
+	vsed -i configure -e 's/ -l:libf2c\.a//'
+	cd "${wrksrc}/vosk-api-${version}/src"
+	vsed -i Makefile -e '/libblas\.a \\/ {N; s/ \\.*//}'
+	vsed -i Makefile -e 's/ -lf2c//'
```
I think this should also be a patch.

JohnGebbie replied:
I didn't make this into patches but I improved the second last vsed, making it explicit and less fragile.
I can still make it into patches if you'd prefer that.

JohnGebbie replied:
I have now made them patches and put links to my pull requests in the template

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,79 @@
+	vsed -i Makefile -e '/libblas\.a \\/ {N; s/ \\.*//}'
+	vsed -i Makefile -e 's/ -lf2c//'
+
+	cd "${wrksrc}/kaldi-${_kaldi_commit}/tools"
+	mkdir -p OpenBLAS/install/lib
+	ln -sf "${XBPS_CROSS_BASE}/usr/include/openblas" OpenBLAS/install/include
+	ln -sf "${XBPS_CROSS_BASE}/usr/lib/"* OpenBLAS/install/lib
+	mkdir -p openfst/include openfst/lib
+	ln -sf "${XBPS_CROSS_BASE}/usr/include/fst" openfst/include
+	ln -sf "${XBPS_CROSS_BASE}/usr/lib/"* openfst/lib
```
Would it be possible to instead add  -I and -L to the build system?

JohnGebbie replied:
I found the kaldi ./configure has --openblas-root and --fst-root options but I think I'd still need have the headers and objects in the same directory.
But I also found vosk also uses these kaldi directories by default and simplified the build args.

JohnGebbie replied:
I still think this is the best way as vosk and kaldi book look to these directories by default and it keeps things simple

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,79 @@
+}
+
+do_install() {
+	cd "${wrksrc}/vosk-api-${version}/src"
```
This is $build_wrksrc, so I think this cd shouldn't be necessary, you should already be there. I might be wrong.

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,79 @@
+vosk-model-small-en-us_package() {
+	short_desc="Lightweight english language model"
+	pkg_install() {
+		vmkdir usr/share/vosk-models
+		mv "${wrksrc}/vosk-model-small-en-us-0.15" "${PKGDESTDIR}/usr/share/vosk-models/small-en-us"
```
I would prefer this to be in `do_install()` and this subpackage's `pkg_install()` would just `vmove` it

paper42 reviewed (commented):
Sorry for the delays, I am definitely interested in merging this and numen.

Could you try upstreaming/reporting issues that you are working around in the template (sse on cross in openfst, libf2c) and link the issues/PRs in the templates?

JohnGebbie commented:
Still ready to merge. I made the vseds into patches and put links to the pull requests in the template as requested.

JohnGebbie commented:
Hold the phone, might be able to simplify and remove a patch.

JohnGebbie commented:
There we go, simplified it slightly by using --mathlib=OPENBLAS which works with lapack out of the box.

JohnGebbie commented:
I reverted back to my pull request patches as they're still an improvement.

JohnGebbie commented:
ready to merge

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+license="Apache-2.0"
+homepage="https://alphacephei.com/vosk"
+distfiles="https://github.com/alphacep/vosk-api/archive/v${version}/vosk-api-${version}.tar.gz
+https://github.com/alphacep/kaldi/archive/${_kaldi_commit}.tar.gz
+https://alphacephei.com/vosk/models/vosk-model-small-en-us-0.15.zip"
```
```suggestion
 https://github.com/alphacep/kaldi/archive/${_kaldi_commit}.tar.gz
 https://alphacephei.com/vosk/models/vosk-model-small-en-us-0.15.zip"
```

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+https://github.com/alphacep/kaldi/archive/${_kaldi_commit}.tar.gz
+https://alphacephei.com/vosk/models/vosk-model-small-en-us-0.15.zip"
+checksum="50fa82bfbfe3fc5be868237c84ac180eb55da0b65b1803bf022899673afda024
+39cadd5a351a5513e3e4e1f71ac4b4f018a592a856a55f1d32c159e22e799f64
+30f26242c4eb449f948e42cb302dd7a686cb29a3423a8367f99ff41780942498"
```
```suggestion
 39cadd5a351a5513e3e4e1f71ac4b4f018a592a856a55f1d32c159e22e799f64
 30f26242c4eb449f948e42cb302dd7a686cb29a3423a8367f99ff41780942498"
```

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+revision=1
+create_wrksrc=yes
+build_wrksrc="vosk-api-${version}/src"
+_kaldi_commit=ecb4b47159501ef0d67888633e65b8fc0c1f79d1
```
Could you add a comment about where we can find this commit?

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+30f26242c4eb449f948e42cb302dd7a686cb29a3423a8367f99ff41780942498"
+shlib_provides="libvosk.so"
+
+do_patch() {
+	# https://github.com/alphacep/kaldi/pull/5
+	patch -Np1 -d "${wrksrc}/kaldi-${_kaldi_commit}" < "${PATCHESDIR}/kaldi-lapack.patch"
+	patch -Np1 -d "${wrksrc}/kaldi-${_kaldi_commit}" < "${PATCHESDIR}/kaldi-to-openblas-0.3.21.patch"
+	patch -Np1 -d "${wrksrc}/kaldi-${_kaldi_commit}" < "${PATCHESDIR}/kaldi-to-openfst-1.8.2.patch"
+	# https://github.com/alphacep/vosk-api/pull/1215
+	patch -Np1 -d "${wrksrc}/vosk-api-${version}" < "${PATCHESDIR}/vosk-lapack.patch"
+}
```
This can be replaced by `patch_args="-Np1 -d ${pkgname}-${version}"` and `patches/kaldi-*.patch.args` files containing `-Np1 -d kaldi`. The kaldi directory currently contains version in its name like I suggest in a comment in `post_extract`. Then patches/series can be removed.

Not overriding do_patch is more consistent with other packages and gets rid of one unnecessary step when adding a patch.
Also a nitpick that's not important if we modify it this way, we usually try to follow keep the order of build steps in sync with order of functions, patch runs after extract, so do_patch should be after do_extract.

JohnGebbie replied:
Thank you! that is better.

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+fi
+
+post_extract() {
+	cd "${wrksrc}/kaldi-${_kaldi_commit}/tools"
```
so we can easily use patch.args
```suggestion
	mv kaldi-${_kaldi_commit} kaldi
	cd "${wrksrc}/kaldi/tools"
```

paper42 commented on srcpkgs/vosk-api/template:
```diff
@@ -0,0 +1,78 @@
+shlib_provides="libvosk.so"
+patch_args="-Np1 -d ${pkgname}-${version}"
+
+# srcpkgs/vosk-api/patches/kaldi-lapack.patch:
+#  https://github.com/alphacep/kaldi/pull/5
+# srcpkgs/vosk-api/patches/vosk-lapack.patch:
+#  https://github.com/alphacep/vosk-api/pull/1215
```
these comments can be put before the first lines of the patches

JohnGebbie commented:
Tested numen using this and all well

JohnGebbie commented:
Thank you! :)))))!
